문자열 사칙 연산 계산기 구현

felix-mr opened the pull request:
### 고민거리
- ExpressionCalculator가 Parser와 Calculator를 주입 받고 두 객체를 통해서 계산을 해주는 방식으로 구현했습니다. 외부에서 주입받은 Parser의 parse 메서드는 ExpressionCalculator 내부에서는 어떻게 동작하는지 모르기 때문에 의존성이 분리되었다고 생각합니다. 여기서 excute 메서드 안에서 parser와 calculator를 같이 사용해주고 있는데 이것은 잘 분리된 것인지 아닌지... 더 나눌 수 있을지 궁금합니다..

일단 테스트 자체가 맞게 하는건지 확신이 서지 않습니다.. 테스트 관련해서 혹시 조언해주고 싶은 부분이 있으신지도 궁금합니다.

felix-mr committed:
SplitStringTest 파일 추가

felix-mr committed:
SplitStringTest 유닛 테스트 작성

felix-mr committed:
SetTests 파일 생성

felix-mr committed:
유닛 테스트 작성

- numbers count 테스트

felix-mr committed:
네이밍 변경

- SplitStringTest -> SplitStringTests

felix-mr committed:
CalculatorTests 파일 추가

felix-mr committed:
Calculator 유닛 테스트

- add, substract, divide, multiply 유닛 테스트

felix-mr committed:
Calculator 리팩토링

- caculate 메서드로 통합

felix-mr committed:
Calculator 리팩토링

- Calculatable을 채택할 수 있도록 수정

felix-mr committed:
Revert "Calculator 리팩토링"

This reverts commit c6bc52d2a1c999534c8bbb52e3f297108035b654.

felix-mr committed:
ExpressionParserTests 파일 추가

felix-mr committed:
ExpressionParser 유닛테스트 작성

felix-mr committed:
Calculator operator(_:)메서드 추가

- string값을 통해 operator를 반환받는 메서드 추가

felix-mr committed:
ExpressionCalculator 파일 추가

felix-mr committed:
ExpressionCalculator 유닛 테스트 추가

- Error 케이스 테스트 추가
- String 값을 통한 계산 테스트 추가

felix-mr committed:
Merge pull request #2 from felix-mr/string-calculator

문자열 사칙 연산 계산기 구현

felix-mr committed:
Merge branch 'felix-mr' into felix-mr

wonhee009 commented on RacingCarTest/CalculatorTests.swift:
```diff
@@ -0,0 +1,107 @@
+    }
+}
+
+class Calculator {
```
`Calculator`가 `CalculatorTests.swift` 안에 정의되어 있네요.
둘을 분리할 필요는 없을까요?🤔

felix-mr replied:
그렇네요.. 테스트만 생각하다 보니까 파일 분리를 하지 못했습니다 ㅠㅠ
앞으로 신경써서 작성하도록 하겠습니다! :)

wonhee009 commented on RacingCarTest/CalculatorTests.swift:
```diff
@@ -0,0 +1,107 @@
+class Calculator {
+    
+    enum `Operator` {
+        
```
줄바꿈이 하나 더 들어간거 같습니다!

felix-mr replied:
음.. class, struct, enum, protocol 등에 대한 정의를 할 때는 개인적으로 한칸을 띄어쓰는 컨벤션을 사용하는데요...

이유는 한 칸을 띄웠을 때 좀 더 답답하지 않아 보이는? 느낌을 받아서 그렇습니다.

해당 컨벤션을 유지하면 문제가 있을까요..?

wonhee009 replied:
앗! 아녀 문제 없습니다! 컨벤션인지 모르고 잘못 들어갔다고 생각했어요😀

wonhee009 commented on RacingCarTest/CalculatorTests.swift:
```diff
@@ -0,0 +1,107 @@
+        case multiply
+    }
+    
+    func `operator`(_ operatorString: String) throws -> Operator? {
```
문자열을 Operator case로 변환하는 메서드 같은데요.
`func operator()`로는 의미 파악이 힘든거 같아요.
좀 더 명확한 이름으로 바꿔보는건 어떨까요?

felix-mr replied:
그렇네요. make나 create를 붙여서 좀 더 명확할 수 있도록 변경하겠습니다!

wonhee009 commented on RacingCarTest/CalculatorTests.swift:
```diff
@@ -0,0 +1,107 @@
+        case "-": return .subtract
+        case "/": return .divide
+        case "*": return .multiply
+        default: return nil
```
`func operator() throws`인데 에러를 던져주는 곳이 없네요?🤔

felix-mr replied:
그렇네요 디폴트의 경우 에러를 던질 수 있도록 수정하겠습니다!

wonhee009 commented on RacingCarTest/CalculatorTests.swift:
```diff
@@ -0,0 +1,107 @@
+        }
+    }
+    
+    private func add(_ left: Int, to right: Int) -> Int {
+        return left + right
+    }
+    
+    private func subtract(_ left: Int, from right: Int) -> Int {
+        return right - left
+    }
+    
+    private func divide(_ left: Int, into right: Int) -> Int {
+        return left / right
+    }
+    
+    private func multiply(_ left: Int, by right: Int) -> Int {
+        return left * right
+    }
```
left는 `_`이고 right는 `to`, `from`, `into`, `by`로 다양하게 정의했네요.
기준이 있을까요?

felix-mr replied:
메소드의 이름과 인자레이블을 작성할 때 문장처럼 읽히도록 작성하고 싶었습니다!

여기서 고민했던게 subtract의 경우 영문법 상으로 `subtract A from B`의 경우 다른 메서드들과는 다르게 순서가 반대가 되더라고요...

그래도 문법상 맞게 실제 말처럼 읽히는게 우선이라고 생각해서 정의했습니다!

wonhee009 commented on RacingCarTest/ExpressionCalculatorTests.swift:
```diff
@@ -0,0 +1,133 @@
+class ExpressionCalculator {
+
+    enum ExpressionError: Error {
+
```
줄바꿈이 하나 더 들어간거 같습니다!

wonhee009 commented on RacingCarTest/ExpressionParserTests.swift:
```diff
@@ -0,0 +1,65 @@
+    
+    func parse(expression: String) throws -> [String] {
+        guard !expression.isEmpty
+        else { throw ParsingError.emptyInput }
```
`else`를 한줄 아래에 쓴거는 의도한 것인가요?

felix-mr replied:
네네 의도한 부분이 맞습니다!

근데 뭔가 일정하지 않게 사용하고 있었네요..ㅠㅠ 다시 한번 체크해서 통일시키도록 하겠습니다!

wonhee009 commented on RacingCarTest/ExpressionCalculatorTests.swift:
```diff
@@ -0,0 +1,133 @@
+    }
+    
+    func calculateRecursively(expression: [String]) throws -> Int {
+        guard !expression.isEmpty else { throw ExpressionError.invalidExpression }
+        guard expression.count > 1 else {
+            
+            if let lastExpression = expression.first,
+               let lastValue = Int(lastExpression) {
+                return lastValue
+            }
+            
+            throw ExpressionError.invalidOperand
+        }
```
expression validation하는 부분을 따로 떼어낼 수 있을거 같아요.

felix-mr replied:
수정하겠습니다! :)

wonhee009 commented on RacingCarTest/ExpressionCalculatorTests.swift:
```diff
@@ -0,0 +1,133 @@
+            throw ExpressionError.invalidExpression
+        }
+        
+        return calculator.calculate(left: left, right: right, operator: `operator`)
```
재귀를 사용해서 계산을 했는데요.
expression이 길다면 for문과 재귀를 사용할때 어떤 차이가 있을까요?

felix-mr replied:
으음 expression이 길어지게 되면 스택이 중첩되면 문제가 생길 것 같습니다.

여기서는 코드를 좀 짧게 줄여보고자 재귀를 사용했는데 for문을 사용하는게 더 나을 것 같네요!

wonhee009 commented on RacingCarTest/ExpressionCalculatorTests.swift:
```diff
@@ -0,0 +1,133 @@
+        self.expressionCalculator = nil
+    }
+    
+    func test_expression_calculator_calculate() {
+        // Arrange
+        let expression = "2 + 3 * 4 / 2"
+        
+        // Act
+        let result = try? expressionCalculator.excute(expression: expression)
+        
+        // Assert
+        XCTAssertEqual(10, result)
+    }
```
given-when-then에 대해서 알아보시는 것도 좋을거 같습니다!😀

felix-mr replied:
이 부분에 대해서 좀 알아봤는데..

[reference](https://softwareengineering.stackexchange.com/questions/308160/differences-between-given-when-then-gwt-and-arrange-act-assert-aaa)

제가 찾아보기로는 큰 차이가 없을 것 같은데

혹시 제가 알아보지 못한 더 큰 차별성이 있을까요...?

wonhee009 replied:
앗 아녀 차별성이라기 보다는 주석 부분이 익숙지 않아서 말씀드렸습니다! 사실 현재 테스트 진행하는 순서랑 별 차이는 없습니다😀

wonhee009 reviewed (changes requested):
@felix-mr 수고하셨습니다!🎉
전체적으로 역할이나 의존성에 대해서 고민을 많이 하신거 같습니다...!
현재 Test 관련 파일 안에 test에서 사용하고 있는 클래스들이 정의되어 있는거 같은데 다른 파일로 분리하는건 어떨지 먼저 의견 드려봅니다!😀

felix-mr committed:
코드리뷰 반영

- https://github.com/next-step/swift-racingcar/pull/17/files/9494f418d30b66d768ac62a8ee2759be0aad6395#r741570218

felix-mr committed:
코드리뷰 반영

- https://github.com/next-step/swift-racingcar/pull/17/files/9494f418d30b66d768ac62a8ee2759be0aad6395#r741570081

felix-mr committed:
코드리뷰 반영

- https://github.com/next-step/swift-racingcar/pull/17/files/9494f418d30b66d768ac62a8ee2759be0aad6395#r741569619

felix-mr committed:
코드리뷰 반영

- https://github.com/next-step/swift-racingcar/pull/17/files/9494f418d30b66d768ac62a8ee2759be0aad6395#r741576723

felix-mr committed:
Merge pull request #3 from felix-mr/string-calculator

String calculator

wonhee009 reviewed (approved):
@felix-mr 수고하셨습니다!🎉
몇몇 리뷰에는 댓글 달아놨습니다!🙏

felix-mr commented:
> @felix-mr 수고하셨습니다!🎉 몇몇 리뷰에는 댓글 달아놨습니다!🙏

확인했습니다! 감사합니다 :)
